src/extension_commands.ml

```diff
@@ -643,27 +628,39 @@ module Search_by_type = struct
                   text_editor
                   client
               in
-              let* type_picker = display_search_results type_search_results in
-              match type_picker with
-              | Some text ->
-                let+ type_inserted =
-                  TextEditor.edit
-                    text_editor
-                    ~callback:(fun ~editBuilder ->
-                      TextEditorEdit.insert
-                        editBuilder
-                        ~location:position
-                        ~value:text)
-                    ()
+              match type_search_results with
+              | [] ->
+                show_message `Info "Empty type/polarity search results";
+                search_by_type ~query ()
```

Comment:
There is two parts in the large `search_by_type` function:
1. Check that the `text_editor` and  the server are compatible with the feature
2. Use UI element to interact with the users and provide feedback

When the results are empty, or when we restart the workflow for whatever reason, we don't need to redo the checks already made at step 1. It would make the code clearer, and the behavior simpler, to extract phase 2 to a separate function that will be the one looping until the user is satisfied.

The part to extract would be:
```
            let position =
              TextEditor.selection text_editor |> Selection.active
            in
            let* query_input = get_query_input ~previous_query:query () in
            match query_input with
            | Some query -> (
              let* type_search_results =
                get_search_results
                  ~query
                  ~with_doc:true
                  ~limit:100
                  ~position
                  text_editor
                  client
              in
              match type_search_results with
              | [] ->
                show_message `Info "Empty type/polarity search results";
                search_by_type ~query ()
              | results -> (
                let* type_picker =
                  display_search_results
                    (remove_duplicates
                       ~cmp:(fun
                           (left :
                             Custom_requests.Type_search.type_search_result)
                           right
                         ->
                         String.equal left.name right.name
                         && left.cost = right.cost)
                       results)
                in
                match type_picker with
                | Some text ->
                  let+ type_inserted =
                    TextEditor.edit
                      text_editor
                      ~callback:(fun ~editBuilder ->
                        TextEditorEdit.insert
                          editBuilder
                          ~location:position
                          ~value:text)
                      ()
                  in
                  if not type_inserted then
                    show_message `Error "Unable to insert %s" text
                | None -> Promise.return ()))
```

I don't have a good name for that function right now. Something like `start_user_interaction`,  or `wait_for_user_input_and_display_results`.